### PR TITLE
#51177: Add an experimental Quasar port of Llama 3.2 1B

### LINKED ISSUE #51177: Port new version of llama to quasar
PR https://github.com/tenstorrent/tt-metal/pull/49292 added a new version of llama

Need to create a new variant:

llama32_1b is at models/common/models/llama32_1b
Would need to create a copy of that and all of its dependencies
models/common/models/executor.py
models/common/models/generator.py
models/common/modules
models/common/sampling
models/common/tests/demos/llama32_1b

Copy should be in
models/experimental/
Potential names:
models/experimental/tt_transformers_qsr
models/experimental/llama32_1b_quasar

### PR DESCRIPTION
**Pull request by bbradelTT**
### PR Category

Feature

### Summary

- Closes #51177 
- Updates codeowners
- Copies over lama32_1b at models/common/models/llama32_1b and all of its dependencies, from various locations, e.g.
models/common/models/executor.py
models/common/models/generator.py
models/common/modules
models/common/sampling
models/common/tests/demos/llama32_1b

### Notes for reviewers

This is a mostly self-contained copy (nothing calls it although CI jobs will be added later, although it does depend on some files outside of the copy), with some refactoring to address AI comments, to use pytest.raises, and to use the files copied over for imports instead of the originals.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:bbradel-51177_llamaq)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:bbradel-51177_llamaq)

### CODE REVIEW

File: models/experimental/llama32_1b_quasar/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+    Environment variables:
+        TT_DEVICE_LOCK_PATH: Override lock file path (default: /tmp/tt_device.lock)
+        TT_DEVICE_LOCK_TIMEOUT: Override timeout in seconds (default: 300)
+    """
+    lock_dir = os.path.dirname(lock_path)
+    if lock_dir and not os.path.exists(lock_dir):
+        os.makedirs(lock_dir, exist_ok=True)
+
+    lock_file = open(lock_path, "a+")  # open the file in append mode to avoid truncation race condition among processes
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/tests/conftest.py:75]: Updating.

File: models/experimental/llama32_1b_quasar/sampling/generator.py [hunk 2]
```diff
@@ -0,0 +1,932 @@
...
+            self._seed_mapper = None
+
+    def _next_unseeded_rng_seed(self) -> int:
+        return secrets.randbits(64)
+
+    def _next_unseeded_device_seed(self) -> int:
+        return secrets.randbelow(DEVICE_SEED_MAX) + 1
+
+    def _next_device_seed_from_rng(self, rng: random.Random) -> int:
+        return rng.randint(1, DEVICE_SEED_MAX)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/sampling/generator.py:652]: This is used in a test. That's ok.

File: models/experimental/llama32_1b_quasar/tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        try:
+            if req_shape != parent_shape:
+                parent_device = ttnn.open_mesh_device(mesh_shape=ttnn.MeshShape(parent_shape), **updated_params)
+                submesh_device = parent_device.create_submesh(ttnn.MeshShape(req_shape))
+                yield submesh_device
+            else:
+                parent_device = ttnn.open_mesh_device(mesh_shape=ttnn.MeshShape(parent_shape), **updated_params)
+                yield parent_device
+        except Exception as e:
+            pytest.skip(f"{__file__}: Mesh device unavailable or unsupported for this configuration: {e}")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/tests/conftest.py]: Updating.

File: models/experimental/llama32_1b_quasar/modules/lazy_weight.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        Generate a unique fingerprint from the conversion configuration.
+
+        This ensures cache invalidation when config changes.
+        """
+        parts = []
+
+        # source
+        # todo)) add better fingerprinting for the source tensor to enable cache invalidation when the source tensor changes; for now, use shape.
+        #        maybe a sampled hash (on, say, first N bytes) would be a good solution --> needs testing! --> test_rope_1d.py is a good test case with differring rope_theta values
+        parts.append(f"srcshape_{'_'.join(str(dim) for dim in self.source.shape)}")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/modules/lazy_weight.py:190]: I need to retain functionality of the original. Keeping this as is.

File: models/experimental/llama32_1b_quasar/modules/rope/rope_1d.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+        Returns:
+            [cos, sin] — height-sharded rotation matrices.
+        """
+        self.load_device_weights()
+        cfg = self.config
+
+        # Send to device if needed
+        if rot_idxs.device != cfg.device:
+            rot_idxs = ttnn.to_device(rot_idxs, cfg.device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/modules/rope/rope_1d.py]: Updating

File: models/experimental/llama32_1b_quasar/models/llama32_1b/generator.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        gen_cfg = gen_cfg or Llama32_1BGeneratorConfig()
+        model = Llama32_1BTransformer1D.from_pretrained(
+            mesh_device,
+            gen_cfg.hf_model_id,
+            max_batch_size=gen_cfg.max_batch_size,
+            max_seq_len=gen_cfg.max_seq_len,
+            num_layers=gen_cfg.num_layers,
+            cache_dir=gen_cfg.cache_dir,
+            precision=gen_cfg.precision,
+            executor_mode=True,
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/llama32_1b/generator.py:72]: Keeping this as is to be the same as original.

File: models/experimental/llama32_1b_quasar/utility_functions.py [hunk 7]
```diff
@@ -0,0 +1,1245 @@
...
+    # Combine zeros and random values
+    combined = torch.cat([zeros, random_values])
+
+    # Shuffle the tensor
+    shuffled = combined[torch.randperm(combined.size(0))]
+
+    # Reshape to the desired shape
+    result_tensor = shuffled.view(shape)
+    result_tensor.to(dtype)
+    return result_tensor
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/utility_functions.py]: Updating.

File: models/experimental/llama32_1b_quasar/models/llama32_1b/demo.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        page_table = make_contiguous_page_table(1, ma.max_seq_len, 32)
+        toks = torch.zeros(1, seq_len, dtype=torch.long)
+        toks[0, :4] = torch.tensor([1, 2, 3, 4], dtype=torch.long)
+        logits = ex.prefill_forward(toks, page_table=page_table, kv_cache=kv)
+        assert logits.shape[0] == 1, f"expected batch dim 1, got {logits.shape}"
+        assert (
+            logits.shape[-1] == model.vocab_size
+        ), f"expected last dim == vocab_size={model.vocab_size}, got {logits.shape}"
+    except Exception as e:
+        pytest.skip(f"Executor prefill not runnable: {e}")
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/llama32_1b/demo.py]: Updating.

File: .github/CODEOWNERS [hunk 9]
```diff
@@ -584,6 +584,7 @@ models/experimental/tt_transformers_v2 @gwangTT @yieldthought @uaydonat @tenstor
 models/experimental/tt_symbiote @alnah005 @mbahnasTT @yieldthought @uaydonat @tenstorrent/codeowner-bypass
 models/experimental/retinanet @dvartaniansTT @tenstorrent/cse-developer-ttnn @tenstorrent/codeowner-bypass
 models/experimental/efficientdetd0 @dvartaniansTT @tenstorrent/cse-developer-ttnn @tenstorrent/codeowner-bypass
+models/experimental/llama32_1b_quasar @tenstorrent/metalium-developers-mmfusedreduce @tenstorrent/codeowner-bypass
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply .github/CODEOWNERS:587]: Updating.

File: models/experimental/llama32_1b_quasar/sampling/README.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  prompt/output state, and to run sampling with or without trace capture.
+- `format_sampling_params`: utility that pads/clamps sampling parameters to the
+  hardware-friendly layout expected by `TTSampling`.
+- `LogProbsCalculator`: computes per-token log-probabilities across a sharded
+  vocabulary using numerically stable log-softmax (global max / sum-exp
+  reduction across devices).
+
+## Quick Start
+```python
+from models.common.sampling import SamplingGenerator, format_sampling_params
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/sampling/README.md]: Updating.

File: models/experimental/llama32_1b_quasar/models/executor.py [hunk 11]
```diff
@@ -0,0 +1,3190 @@
...
+import torch
+from loguru import logger
+
+import ttnn
+from models.experimental.llama32_1b_quasar.models.module_input_validation import (
+    suspend_module_input_validation,
+    validate_module_input_configs,
+)
+from models.experimental.llama32_1b_quasar.sampling.sampling_params import SamplingParams
+from models.experimental.llama32_1b_quasar.tests.demos.cleanup_utils import cleanup_ttnn_value
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/executor.py:39]: Copied over from original. If it is ok there, it should be ok for this copy to be used for specific testing scenarios. Leaving as is.

File: models/experimental/llama32_1b_quasar/models/llama32_1b/model.py [hunk 12]
```diff
@@ -0,0 +1,1319 @@
...
+# =============================================================================
+# Llama32_1BTransformer1D
+# =============================================================================
+
+
+class Llama32_1BTransformer1D(LightweightModule):
+    """TTTv2 Llama 3.2-1B-Instruct transformer.
+
+    Construct via :meth:`from_pretrained`. Sub-modules (``embedding``, ``rope_setup``,
+    ``layers``, ``norm``, ``lm_head``) are pre-built from HF weights with ``LazyWeight``.
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/llama32_1b/model.py:620]: I will need to add this model to a separate job in CI once I have it merged and more tests added to it. That will be for a future PR.

File: models/experimental/llama32_1b_quasar/modules/sampling/penalties_1d.py [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+        )
+        src_tt = ttnn.from_torch(
+            src_host,
+            device=self.config.mesh_device,
+            dtype=ttnn.int32,
+            layout=ttnn.ROW_MAJOR_LAYOUT,
+            mesh_mapper=self._replicate_mapper,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        )
+        self._token_bin_counts_and_mask(new_tokens=prompt_tokens_tt, src=src_tt, mask=params.prompt_mask)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/modules/sampling/penalties_1d.py:229]: Leaving functionality as is since this is in the original.

File: models/experimental/llama32_1b_quasar/models/generator.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+"""
+Llama3Generator — thin vLLM adapter wrapping an executor.
+
+Zero trace state, zero warmup state, zero execution logic.
+Just signature adaptation for TTModelRunner.
+"""
+
+import ttnn
+from models.common.models.llama3_8b.model import EagerLlamaExecutor, Llama3Transformer1D, TracedLlamaExecutor
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/generator.py:12]: This PR is already large. If this becomes an issue later on, I will copy this over as well. I updated the description.

File: models/experimental/llama32_1b_quasar/models/executor.py [hunk 15]
```diff
@@ -0,0 +1,3190 @@
...
+from models.tt_transformers.tt.common import (
+    Mode,
+    copy_host_to_device,
+    get_block_size,
+    get_max_prefill_chunk_size,
+    get_padded_prefill_len,
+    num_blocks_in_seq,
+)
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/models/executor.py:47]: This PR is already large. If this becomes an issue later on, I will copy this over as well. I updated the description.

**Review by mtairum — APPROVED** [review]

File: models/experimental/llama32_1b_quasar/tests/conftest.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        TT_DEVICE_LOCK_TIMEOUT: Override timeout in seconds (default: 300)
+    """
+    # Re-sanitize at the sink: guarantee the path reaching the filesystem is
+    # confined to the trusted temp dir, regardless of what the caller passed.
+    lock_path = _sanitize_lock_path(lock_path)
+    lock_dir = os.path.dirname(lock_path)
+    if lock_dir and not os.path.exists(lock_dir):
+        os.makedirs(lock_dir, exist_ok=True)
+
+    lock_file = open(lock_path, "a+")  # open the file in append mode to avoid truncation race condition among processes
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply models/experimental/llama32_1b_quasar/tests/conftest.py:75]: I'll keep this the same as the copied code.

**Review by williamlyTT — APPROVED** [review]
